Add Product Reviews block

sunyatasattva opened the pull request:
This PR adds a “Product Reviews” block which can be inserted inside the Single Product template, enabling users to display the `single-product-reviews` template provided by their theme or WooCommerce core.

On the editor side, it shows a placeholder to make it easier for the users to visualize their template.

Note: this PR doesn't change the default Single Product template.

Fixes #8856

#### Developers notes

See this discussion for more context on the implementation: https://github.com/woocommerce/woocommerce-blocks/discussions/8792

I have implemented solution (1.) suggested in the OP, because it was the lowest hanging fruit and didn't require modifications to core. I'd wager that solution (2.) would be the best: but we can change the implementation details transparently even after merging the PR and cleaning things up.

Regarding what should be shown, you should be able to override the contents by changing the provided `woocommerce/single_products_reviews.php` template file in your theme folder.

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

#### Editor side

![Screenshot 2023-03-27 at 00 35 57](https://user-images.githubusercontent.com/1847066/227809035-514e8aee-f3ed-40dc-8928-bb496530e094.png)

#### Front-end

![Screenshot 2023-03-27 at 00 37 07](https://user-images.githubusercontent.com/1847066/227809094-fc5376e4-72e4-4b10-89b7-af8288b5603a.png)


### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

0. Have a product with at least one review.
1. Go to Appearance > Editor > Single Product Template.
2. Make sure the “Product Reviews” block appears in the inserter and add it.
3. Make sure a placeholder appears in the editor and save.
4. Go to the front end and make sure that the reviews are correctly shown.

##### Known issues

1. The Product Reviews block stars won't work if there is one or more review blocks inserted before (for example, via the Classic Single Product template, or via the Product Details block). This is because the stars [are added on page load via jQuery](https://github.dev/woocommerce/woocommerce/blob/ca49caabcba84ce9f60a03c6d3534ec14b350b80/plugins/woocommerce/client/legacy/js/frontend/single-product.js#L46) on only a single `#rating` block. We could perhaps consider adding our own JavaScript associated with this block, but this is an edge case (having multiple reviews block) which we might not even want to support, so it's not added to this PR yet.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Added the “Products Review” block

sunyatasattva committed:
Add Product Reviews block

sunyatasattva committed:
Fix PHP linting issues

sunyatasattva committed:
Ignore PHPCS on overriding globals

sunyatasattva committed:
Merge branch 'trunk' into add/8856-product-reviews

nefeline commented on src/BlockTypes/ProductReviews.php:
```diff
@@ -0,0 +1,68 @@
+	 * @return string Rendered block output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		$post_id = $block->context['postId'];
+		$old_global_product = $GLOBALS['product'];
+		$old_global_query = $GLOBALS['wp_query'];
+
+		$new_query = new \WP_Query(
+			array(
+				'feed'         => 1,
+				'p'            => $post_id,
+				'post_type'    => 'product',
+				'withcomments' => 1,
+			)
+		);
+
+		// Ovewriting the globals, because we need it in the template.
+		$GLOBALS['product'] = wc_get_product( $post_id );
+		$GLOBALS['wp_query'] = $new_query; // phpcs:ignore
+
+		wc_get_template( 'single-product-reviews.php' );
+
+		$reviews = ob_get_clean();
+
+		// Resetting the globals to what it was.
+		$GLOBALS['product'] = $old_global_product;
+		$GLOBALS['wp_query'] = $old_global_query; // phpcs:ignore
```
What do you think about following this route?

```suggestion
		ob_start();

		rewind_posts();
		while ( have_posts() ) {
			the_post();
			comments_template();
		}

		$reviews = ob_get_clean();
```

By invoking `comments_template();` instead of `wc_get_template( 'single-product-reviews.php' );`, we:

- Ensure all necessary variables are in place.
- Defer the loading of the `'single-product-reviews.php'` template to the core of Woo: the correct template will still be loaded as soon as the `comment_template` filter is triggered.

Also, we can benefit from `rewind_posts()` for updating the internal pointer for the posts and preventing the conflicts with the Post Content block that we were discussing [over here](https://github.com/woocommerce/woocommerce-blocks/discussions/8792#discussioncomment-5387577).

sunyatasattva replied:
This looks good, thanks @nefeline! I personally remain with my doubts considering this approach. But you seem to hold a stronger opinion on the matter, so I'll follow your lead and change the code to your suggestion! 😊

nefeline replied:
> This looks good, thanks @nefeline! I personally remain with my doubts considering this approach. But you seem to hold a stronger opinion on the matter, so I'll follow your lead and change the code to your suggestion! 😊

Thanks! What do you think about asking for some guidance from the GB team on this matter? All of my points of view [shared in our dicussion](https://github.com/woocommerce/woocommerce-blocks/discussions/8792#discussioncomment-5387577) so far are based on what I saw implemented in other blocks on Gutenberg (post content, featured image, post comments), and I think it's wise for us to follow similar paths as they did, but if they have a different recommendation to offer it would be important to know.

sunyatasattva replied:
I'm totally fine with proceeding with this direction, especially because it is consistent with what Gutenberg does. The discussion could stay open in case anyone wants to share their opinion (Nik from Gutenberg, for instance, was also suggesting the second point), but I feel the need to elaborate any further, personally: you make good points and, at the end of the day, these are implementation details that we can easily change without causing breaking changes if we ever feel the need to, so I'm totally happy to proceed with this, if you are satisfied as well :)

sunyatasattva committed:
Change template rendering logic

Use `rewind_posts` plus `comments_template` instead
of overriding the global variables.

Co-authored-by: Patricia Hillebrandt <patriciahillebrandt@gmail.com>

sunyatasattva committed:
Merge branch 'trunk' into add/8856-product-reviews

nefeline reviewed (approved):
Great work 🎉 !

I'm preemptively approving, but have only one suggestion to make: for all other blocks we created for the Single Product template, we [disabled them for post and page editors](https://github.com/woocommerce/woocommerce-blocks/blob/6d287b77359ba9774d82b3cb76e914b00280f67e/src/BlockTypesController.php#L261-L267): I would recommend doing the same here.

Test results:

> Have a product with at least one review.
> Go to Appearance > Editor > Single Product Template.
> Make sure the “Product Reviews” block appears in the inserter and add it.
> Make sure a placeholder appears in the editor and save.

✅ Confirmed the placeholder is displayed in the editor.

> Go to the front end and make sure that the reviews are correctly shown as per the provided single_products_reviews.php template file.

✅ Confirmed the reviews are correctly shown on the FE and new reviews can be successfully submitted.

> Known issues
> The Product Reviews block stars won't work if there is one or more review blocks inserted before (for example, via the Classic Single Product template, or via the Product Details block). This is because the stars [are added on page load via jQuery](https://github.dev/woocommerce/woocommerce/blob/ca49caabcba84ce9f60a03c6d3534ec14b350b80/plugins/woocommerce/client/legacy/js/frontend/single-product.js#L46) on only a single #rating block. We could perhaps consider adding our own JavaScript associated with this block, but this is an edge case (having multiple reviews block) which we might not even want to support, so it's not added to this PR yet.

FYI: we already have an open issue for this: https://github.com/woocommerce/woocommerce-blocks/issues/8420 & a "hacky" workaround available. I agree that Ideally, we should have a more robust solution (& possibly upstream it) if we want to support this use case for sure.

sunyatasattva commented:
Excellent @nefeline ! Thank you so much!

> I'm preemptively approving, but have only one suggestion to make: for all other blocks we created for the Single Product template, we [disabled them for post and page editors](https://github.com/woocommerce/woocommerce-blocks/blob/6d287b77359ba9774d82b3cb76e914b00280f67e/src/BlockTypesController.php#L261-L267): I would recommend doing the same here.

I thought [this code](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/c125e5655bf93aaf109267d285d01300dcdf91e8/assets/js/atomic/utils/register-block-single-product-template.ts#L7) took care of that, is that incorrect? Do we need both? 🤔

nefeline commented:
> I thought [this code](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/c125e5655bf93aaf109267d285d01300dcdf91e8/assets/js/atomic/utils/register-block-single-product-template.ts#L7) took care of that, is that incorrect? Do we need both? 🤔

Hm, maybe you are right! `registerBlockSingleProductTemplate` was added later on. If that's the case, feel free to disregard 👍
